### PR TITLE
bpo-23033:  consider wildcard in left most segment only for domain names

### PR DESCRIPTION
**Pull request by daxlab**
https://bugs.python.org/issue23033

### CODE REVIEW

**Comment by alex** [comment]
Looks like there is a relevant failing `ssl` test.

**Comment by daxlab** [comment]
@alex review now.

File: Lib/ssl.py
Comment (alex) [inline Lib/ssl.py]: Nit: Please use a `#` style comment. `"""` is for docstrings.

File: Lib/test/test_ssl.py
Comment (alex) [inline Lib/test/test_ssl.py]: This comment still doesn't fell right, there is a wildcard _in_ the left-most segment. I think it should say something like "only match wildcards when they are the only thing in the left-most segment". What do you think?

**Review by alex — CHANGES_REQUESTED** [review]
I have a few small comments.

Thanks for working on this! This is a great improvement to making `match_hostname` behave more like the rest of the PKI infrastructure.

File: Lib/test/test_ssl.py
Comment (daxlab) [reply Lib/test/test_ssl.py]: Sounds much better, pushing the changes.

**Comment by alex** [comment]
Doc failures are not your fault, https://github.com/sphinx-doc/sphinx/issues/3597 is the cause.

**Review by alex — APPROVED** [review]

**Comment by daxlab** [comment]
@alex  anything else needed from my side ?

**Comment by alex** [comment]
This looks good to me. I'll probably ask for another review before landing. Thanks so much!

**Review by reaperhulk — APPROVED** [review]
Looks good to me. Wildcards should be honored iff they are the leftmost fragment and are the only character in that fragment.

**Comment by tiran** [comment]
You are almost there! Please add a ```versionchanged:: 3.7``` to ```match_hostname``` documentation in ```Doc/library/ssl.rst```.

**Review by tiran — CHANGES_REQUESTED** [review]
versionchanged in documentation is missing.

**Comment by daxlab** [comment]
@tiran version changed docs added. Should that go in `Doc/whatsnew/3.7.rst` also ?

**Review by tiran — APPROVED** [review]
Thank you very much for your contribution, @daxlab. Your work is an excellent base to clean up and simplify the host matching code even further.

@alex the change is backwards incompatible. I've asked Ben and Ned if they are fine with backports to 2.7 and 3.6.

**Comment by daxlab** [comment]
@tiran @alex can we merge this ?

**Comment by tiran** [comment]
@daxlab @alex I'm waiting for confirmation regarding backport to 3.6 and 2.7.

File: Doc/whatsnew/3.7.rst
Comment (berkerpeksag) [inline Doc/whatsnew/3.7.rst]: I'd change :meth:\`match_hostname\` to :meth:\`ssl.match_hostname\`. This would create a link to the `ssl.match_hostname()` documentation.

**Review by berkerpeksag — COMMENTED** [review]
I think we also need to add an entry to `Misc/NEWS`.

File: Doc/whatsnew/3.7.rst
Comment (daxlab) [reply Doc/whatsnew/3.7.rst]: @berkerpeksag thanks for review. I have pushed a fix.

**Comment by daxlab** [comment]
@tiran any updates ?

File: Misc/NEWS
Comment (Mariatta) [inline Misc/NEWS]: Please add "Patch by Mandeep Singh."

**Comment by Mariatta** [comment]
Please add yourself to Misc/ACKS, resolve the conflict and update Misc/NEWS.
Thanks.

File: Misc/NEWS
Comment (daxlab) [reply Misc/NEWS]: done.

**Comment by daxlab** [comment]
@Mariatta changes pushed.

**Comment by daxlab** [comment]
@Mariatta changes pushed.

**Comment by daxlab** [comment]
any update ?

**Comment by the-knights-who-say-ni** [comment]
Hello, and thanks for your contribution!

I'm a bot set up to make sure that the project can legally accept your contribution by verifying you have signed the [PSF contributor agreement](https://www.python.org/psf/contrib/contrib-form/) (CLA).

Unfortunately our records indicate you have not signed the CLA. For legal reasons we need you to sign this before we can look at your contribution. Please follow [the steps outlined in the CPython devguide](https://devguide.python.org/pullrequest/#licensing) to rectify this issue.


Thanks again to your contribution and we look forward to looking at it!

**Comment by Mariatta** [comment]
Thanks @daxlab for your patience with this PR, sorry it's taking a while.
I think this can go to 3.7.
Could you rebase this to make it up to date? We need the news entry in Misc/NEWS.d/, it can be generated using [blurb](https://pypi.org/project/blurb).

**Comment by daxlab** [comment]
@Mariatta will update the PR tonight.

**Comment by daxlab** [comment]
@Mariatta I have rebased the PR and added news entry in Misc/NEWS.d/

**Review by Mariatta — APPROVED** [review]

**Comment by Mariatta** [comment]
Thanks. I made minor changes and added you to Misc/ACKs. Waiting until CI is done.

**Comment by Mariatta** [comment]
Thanks again @daxlab and congrats on your first contribution to CPython :taco:

**Pull request merged**
